api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,600 @@
+	}
+
+	serviceConsumers, err := e.getServiceResourceConsumers(clusterName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	jobTaskConsumers, err := e.getJobTaskResourceConsumers(clusterName)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	resourceConsumers := append(serviceConsumers, jobTaskConsumers...)
+
+	return resourceConsumers, resourceProviders, nil
+}
+
+func (e *EnvironmentScaler) CalculateOptimizedState(clusterName string, resourceConsumers []scaler.ResourceConsumer, resourceProviders []*scaler.ResourceProvider) ([]*scaler.ResourceProvider, []*scaler.ResourceProvider, int, error) {
+	var incompatibilityErrs []error
+	var multiError error
+
+	// calculate for scaling up
+	resourceProviders, errs := e.calculateScaleUp(clusterName, resourceProviders, resourceConsumers)
+	for _, err := range errs {
+		// the IncompatibleConsumerAndProvider error is returned in situations where
+		// a resource consumer is too large to fit into an empty, brand-new resource
+		// provider; this means that the environment's instance size is too small
+		// for the consumer. in such a scenario, we want to perform scaling actions
+		// for other providers and consumers, but also log that there is a problem.
+		if err, ok := err.(*errors.ServerError); ok && err.Code == errors.IncompatibleConsumerAndProvider {
+			incompatibilityErrs = append(incompatibilityErrs, err)
+			continue
+		}
+
+		return nil, nil, 0, err
+	}
+
+	if len(incompatibilityErrs) > 0 {
+		multiError = errors.New(errors.IncompatibleConsumerAndProvider, errors.MultiError(incompatibilityErrs))
+	}
+
+	// calculate for scaling down
+	unusedProviders := e.calculateScaleDown(clusterName, resourceProviders)
+	desiredScale := len(resourceProviders) - len(unusedProviders)
+
+	return resourceProviders, unusedProviders, desiredScale, multiError
+}
+
+func (e *EnvironmentScaler) ScaleToState(clusterName string, desiredScale int, unusedProviders []*scaler.ResourceProvider) error {
+	asg, err := readASG(e.Client.AutoScaling, clusterName)
+	if err != nil {
+		return err
+	}
+
+	currentCapacity := int(aws.Int64Value(asg.DesiredCapacity))
+
+	log.Printf("[DEBUG] [EnvironmentScaler] Attempting to scale environment '%s' from current scale of '%d' to desired scale of '%d'.", clusterName, currentCapacity, desiredScale)
+
+	asgName := aws.StringValue(asg.AutoScalingGroupName)
+	desiredScale64 := int64(desiredScale)
+
+	minCapacity := int(aws.Int64Value(asg.MinSize))
+	maxCapacity := int(aws.Int64Value(asg.MaxSize))
+
+	// sample error returned when desired < min || desired > max:
+	//
+	// 2017/12/14 23:04:39 [ERROR] [Worker] [3]: Failed to run job 1513321478660496503: ValidationError: Desired capacity:1 must be between the specified min size:2 and max size:2
+	//         status code: 400, request id: ed313128-e165-11e7-ab64-c9c2b806a2ed
+	if err := updateASG(e.Client.AutoScaling, asgName, nil, nil, &desiredScale64); err != nil {
+		if err, ok := err.(awserr.Error); ok && err.Code() == "ValidationError" {
```

Comment:
It isn't clear where this error would be returned from?

Reply:
The idea is that when `desiredScale > maxScale` or `desiredScale < minScale`, we don't want to treat that as an error. AWS will return an error about how that's an invalid request, and we want to log that such an event happened in our `[DEBUG]` output, but the desired behavior for Layer0 is to respect the min/max limits imposed by a user.

Reply:
@sesh-kebab I didn't catch the edit earlier, but I ran into this same problem. My solution was to cause a request to be sent to AWS in which `desiredScale < minScale` and inspect what came back in a local API. `ValidationError` is what came back, and that's what I've based the error handling off of.